pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/BaseBooleanAggregateFunction.java

```diff
@@ -0,0 +1,249 @@
+ * under the License.
+ */
+
+package org.apache.pinot.core.query.aggregation.function;
+
+import java.util.Map;
+import org.apache.pinot.common.request.context.ExpressionContext;
+import org.apache.pinot.common.utils.DataSchema;
+import org.apache.pinot.common.utils.PinotDataType;
+import org.apache.pinot.core.common.BlockValSet;
+import org.apache.pinot.core.query.aggregation.AggregationResultHolder;
+import org.apache.pinot.core.query.aggregation.LongAggregationResultHolder;
+import org.apache.pinot.core.query.aggregation.ObjectAggregationResultHolder;
+import org.apache.pinot.core.query.aggregation.groupby.GroupByResultHolder;
+import org.apache.pinot.core.query.aggregation.groupby.LongGroupByResultHolder;
+import org.apache.pinot.core.query.aggregation.groupby.ObjectGroupByResultHolder;
+import org.apache.pinot.spi.data.FieldSpec;
+import org.roaringbitmap.RoaringBitmap;
+
+
+public abstract class BaseBooleanAggregateFunction extends BaseSingleInputAggregationFunction<Integer, Integer> {
+
+  private final BooleanMerge _merger;
+  private final boolean _nullHandlingEnabled;
+
+  protected enum BooleanMerge {
+    AND {
+      @Override
+      long merge(long left, int right) {
+        return left * right;
+      }
+
+      @Override
+      boolean isTerminal(long agg) {
+        return agg == 0;
+      }
+
+      @Override
+      long getDefaultValue() {
+        return 1;
+      }
+    },
+    OR {
+      @Override
+      long merge(long left, int right) {
+        return left + right;
+      }
+
+      @Override
+      boolean isTerminal(long agg) {
+        return agg > 0;
+      }
+
+      @Override
+      long getDefaultValue() {
+        return 0;
+      }
+    };
+
+    abstract long merge(long left, int right);
+
+    abstract boolean isTerminal(long agg);
+
+    abstract long getDefaultValue();
+  }
+
+  protected BaseBooleanAggregateFunction(ExpressionContext expression, boolean nullHandlingEnabled,
+      BooleanMerge merger) {
+    super(expression);
+    _nullHandlingEnabled = nullHandlingEnabled;
+    _merger = merger;
+  }
+
+  @Override
+  public AggregationResultHolder createAggregationResultHolder() {
+    return _nullHandlingEnabled
+        ? new ObjectAggregationResultHolder()
+        : new LongAggregationResultHolder(_merger.getDefaultValue());
+  }
+
+  @Override
+  public GroupByResultHolder createGroupByResultHolder(int initialCapacity, int maxCapacity) {
+    return _nullHandlingEnabled
+        ? new ObjectGroupByResultHolder(initialCapacity, maxCapacity)
+        : new LongGroupByResultHolder(initialCapacity, maxCapacity, _merger.getDefaultValue());
+  }
+
+  @Override
+  public void aggregate(int length, AggregationResultHolder aggregationResultHolder,
+      Map<ExpressionContext, BlockValSet> blockValSetMap) {
+    BlockValSet blockValSet = blockValSetMap.get(_expression);
+
+    if (blockValSet.getValueType().getStoredType() != FieldSpec.DataType.INT) {
```

Comment:
Do we want to support it on `INT` column? If not, we should explicitly check the value type instead of the stored type. The current logic won't work well when the input value is not 0/1.